### PR TITLE
CDPSDX-3525: Added re-registration of cluster proxy config to reattach process for DL resize recovery.

### PR DESCRIPTION
**Pull request by sxxgrc**
Jira: https://jira.cloudera.com/browse/CDPSDX-3525

**Original Problem**:

- DL resize recovery was timing out due to not being able to communicate with the cluster proxy service
- We determined that this was probably due to the configuration for the DL's stack being deleted after the resized DL is deleted during recovery (the CRN is retained throughout the entire resize process)

**Code Change Objective**:

- Add a step in the reattach logic (used during resize recovery after deletion of the resized DL) which re-registers the cluster proxy configuration for the DL's stack
- This effectively required the creation of a new set of methods to get this to work from the DL code since the cluster proxy service is in the _core_ code

Just to reiterate, the entire point of this change is simply to call the **reRegister** method in the cluster proxy service. We figured the best way to do this in order to follow the pre-existing standard was to invoke a new flow to do this.

Tests done:

- Ensure local version of my code changes for this with CCM V2 does not have the original issue 
- Ensure cluster repair, which runs the upscale flow I modified within my changes, is still successful 
- Ensure reattach after failure in detach (which will now call the new cluster proxy method) is still successful 

Please let me know if you have any questions.

### CODE REVIEW

**Review by RokolyA94 — APPROVED** [review]

File: core-api/src/main/java/com/sequenceiq/cloudbreak/doc/OperationDescriptions.java [hunk 1]
```diff
@@ -85,6 +85,7 @@ public static class StackOpDescription {
         public static final String CHANGE_IMAGE_CATALOG = "Changes image catalog of the cluster";
         public static final String RANGER_RAZ_ENABLED = "Determines if Ranger Raz is present in the cluster.";
         public static final String GENERATE_IMAGE_CATALOG = "Generates an image catalog that only contains the currently used image for creating instances";
+        public static final String RE_REGISTER_CLUSTER_PROXY_CONFIG = "Re-registers the cluster proxy config for the cluster if possible.";
```
Comment (kkalvagadda1) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/doc/OperationDescriptions.java:88]: How about "Re-registers the cluster proxy config for the cluster if needed"?

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/clusterproxy/reregister/ClusterProxyReRegistrationEvent.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.sequenceiq.cloudbreak.core.flow2.stack.clusterproxy.reregister;
+
+import com.sequenceiq.cloudbreak.reactor.api.event.orchestration.ClusterProxyReRegistrationResult;
+import com.sequenceiq.flow.core.FlowEvent;
+import com.sequenceiq.flow.event.EventSelectorUtil;
+
+public enum ClusterProxyReRegistrationEvent implements FlowEvent {
+    CLUSTER_PROXY_RE_REGISTRATION_EVENT("CLUSTER_PROXY_RE_REGISTRATION_EVENT"),
+    CLUSTER_PROXY_RE_REGISTRATION_FINISHED_EVENT(EventSelectorUtil.selector(ClusterProxyReRegistrationResult.class)),
+    CLUSTER_PROXY_RE_REGISTRATION_FAILED_EVENT(EventSelectorUtil.failureSelector(ClusterProxyReRegistrationResult.class)),
```
[the remark this answers is not in the document]
Comment (sxxgrc) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/stack/clusterproxy/reregister/ClusterProxyReRegistrationEvent.java:10]: I believe I have fixed this.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java [hunk 3]
```diff
@@ -106,19 +131,33 @@ public void reattachStack(SdxCluster cluster, String originalName) {
...
-        try {
-            sdxAttachDetachUtils.updateExternalDatabase(cluster, detachedCrn);
-            LOGGER.info("Finished reattaching external database for SDX cluster with ID: {}. Now has crn: {}.",
-                    cluster.getId(), cluster.getCrn());
-        } catch (RuntimeException e) {
-            // Undo the stack update to put everything back in its original state.
-            sdxAttachDetachUtils.updateStack(cluster.getClusterName(), detachedName, detachedCrn, true);
-            jobService.schedule(cluster.getId(), SdxClusterJobAdapter.class);
-            throw e;
-        }
```
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:121]: Why are you changing this?

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java [hunk 4]
```diff
@@ -39,30 +54,40 @@ public class SdxAttachService {
...
-                reattachExternalDatabase(reattached, detachedCrn, detachedName);
+                reattachExternalDatabase(reattached, detachedCrn);
             }
             return reattached;
         } catch (Exception e) {
             LOGGER.error("Failed to reattach SDX cluster with CRN '" + detachedCrn +
                     "'. Putting detached cluster back to its original state.", e);
-            sdxAttachDetachUtils.updateClusterNameAndCrn(reattached, detachedName, detachedCrn);
-            reattached.setDetached(true);
-            sdxService.save(reattached);
```
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:65]: Why is this changed?

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/v4/StackV4Controller.java [hunk 5]
```diff
@@ -478,8 +477,6 @@ public FlowIdentifier renewCertificate(Long workspaceId, String name, @Initiator
     public FlowIdentifier renewInternalCertificate(Long workspaceId, @TenantAwareParam String crn) {
         try {
             return stackOperationService.renewInternalCertificate(crn);
-        } catch (FlowNotTriggerableException e) {
-            throw new BadRequestException(e.getMessage());
```
[the remark this answers is not in the document]
Comment (sxxgrc) [reply core/src/main/java/com/sequenceiq/cloudbreak/controller/v4/StackV4Controller.java:482]: It is unrelated. These two lines add nothing to the code as all exceptions are caught right below this and handled the same way.
I simply modified this since I noticed the file giving me a warning about this.
I can revert but don't really see any reason to.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java (hunk 3 above)
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java:121]: Yes this is refactoring. I essentially created a new method which will revert the stack update if an exception occurs during reattach.
The reason I do this is that with the new cluster proxy step we need the same logic so I decided to just have it all centralized in one method.

**Review by lturcsanyi — APPROVED** [review]

**Review by kkalvagadda1 — APPROVED** [review]
Looks good to me.

**Comment by sxxgrc** [comment]
Finished testing and updated the description to describe the tests done. This can now be merged in.
